### PR TITLE
create a function checking username in social auth pipeline

### PR DESCRIPTION
**Pull request by lunika**
## Purpose

When a username is already registered in richie database, python social
auth will prepend the username with a uuid to make it unique. We don't
want this behaviour because API call with edx don't work in this case,
we need the real username. We created a function, added in the python
social auth connection pipeline, responsible to check the username
availability and raise an exception if the same username already exists.

## Proposal

- [x] move backends in social apps
- [x] create a function checking username in social auth pipeline 
- [x] extend social-auth middleware to use a page to display error

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,11 @@ Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
 
+### Added
+
+- Create a function checking username in social auth pipeline
+- extend social-auth middleware to use a page to display error
```
Comment (jmaupetit) [inline CHANGELOG.md:14]:
```suggestion
- Extend social-auth middleware to use a page to display error
```

File: src/richie/apps/social/middleware.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    def get_redirect_uri(self, request, exception):
+        """
+        Check if an error page exists and returns its url. Fallback on parent function otherwise.
+        """
+        if SOCIAL_ERROR_REVERSE_ID:
+            page = Page.objects.get(reverse_id=SOCIAL_ERROR_REVERSE_ID)
+            return page.get_absolute_url(language=get_language_from_request(request))
+
+        return super.get_redirect_uri(request, exception)
```
Comment (sampaccoud) [inline src/richie/apps/social/middleware.py:25]: Maybe more pythonic:
```suggestion
        try:
            page = Page.objects.get(reverse_id=SOCIAL_ERROR_REVERSE_ID)
        except Page.DoesNotExist:
            return super.get_redirect_uri(request, exception)
        else:
            return page.get_absolute_url(language=get_language_from_request(request))
```

File: src/richie/apps/social/pipeline/user.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
+"""Module used by python-social-auth pipeline."""
+from social_core.exceptions import AuthAlreadyAssociated, AuthFailed
+
+USER_FIELDS = ["username", "email"]
+
+
+# pylint: disable=unused-argument,keyword-arg-before-vararg
+def get_username(strategy, details, backend, user=None, *args, **kwargs):
+    """Check if the username already exists. Raised an exception if yes."""
```
Comment (sampaccoud) [inline src/richie/apps/social/pipeline/user.py:9]:
```suggestion
    """Check if the username already exists. Raise an exception if yes."""
```

File: src/richie/apps/social/pipeline/user.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    if "username" not in backend.setting("USER_FIELDS", USER_FIELDS):
+        return None
+
+    storage = strategy.storage
+
+    if user:
+        # The user already exists return its usernme.
```
Comment (sampaccoud) [inline src/richie/apps/social/pipeline/user.py:16]:
```suggestion
        # The user already exists return its username.
```

File: tests/apps/social/test_pipeline_user.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+from social_core.exceptions import AuthAlreadyAssociated, AuthFailed
+
+from richie.apps.social.pipeline.user import USER_FIELDS, get_username
+
+
+class PipelineUserTestCase(TestCase):
+    """Tests for pipeline user."""
+
+    def test_get_username_no_username_in_backend(self):
+        """if username is not un backend settings should return None."""
```
Comment (sampaccoud) [inline tests/apps/social/test_pipeline_user.py:17]:
```suggestion
        """if username is not in backend settings should return None."""
```

File: tests/apps/social/test_pipeline_user.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        backend = mock.Mock()
+        backend.setting.return_value = {}
+        strategy = mock.Mock()
+
+        self.assertIsNone(get_username(strategy, {}, backend))
+
+    def test_get_username_user_connected(self):
+        """The user is already connected, its username is fetch and returned."""
```
Comment (sampaccoud) [inline tests/apps/social/test_pipeline_user.py:25]:
```suggestion
        """The user is already connected, its username is fetched and returned."""
```

File: tests/apps/social/test_pipeline_user.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        """No user connected and the username is already saved. Should throw an exception."""
+        backend = mock.Mock()
+        backend.setting.return_value = USER_FIELDS
+
+        strategy = mock.Mock()
+        strategy.setting.return_value = False
+        strategy.storage.user.user_exists.return_value = True
+
+        with self.assertRaises(AuthAlreadyAssociated):
+            get_username(strategy, {"username": "John Doe"}, backend)
```
Comment (sampaccoud) [inline tests/apps/social/test_pipeline_user.py:89]: I'm not really fan of these type of fully mocked unit tests. It helps for development but they will continue to pass when something changes in python social core... for example if the user_exists method disappears...
I feel creating a real user in db and testing the method with a real strategy and backend is much more security for this kind of override of an existing lib.

Comment (lunika) [reply tests/apps/social/test_pipeline_user.py:89]: Ok I will look how to do this.

Comment (lunika) [reply tests/apps/social/test_pipeline_user.py]: Done !

File: tests/apps/social/test_pipeline_user.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+
+class PipelineUserTestCase(TestCase):
+    """Tests for pipeline user."""
+
+    def setUp(self):
+        """Configure social-auth strategy."""
+        self.request_factory = RequestFactory()
+        self.request = self.request_factory.get("/", data={"x": "1"})
+        SessionMiddleware().process_request(self.request)
```
Comment (sampaccoud) [inline tests/apps/social/test_pipeline_user.py:21]: Curious: 
Can you explain why you need data?
What is the middleware that you  need to run here?

**Review by sampaccoud — APPROVED** [review]
Nice convincing tests ;-)

File: tests/apps/social/test_pipeline_user.py (hunk 8 above)
Comment (lunika) [reply tests/apps/social/test_pipeline_user.py:21]: The `DjangoStrategy` use the session so we have to add it through the `SessionMiddleware`. Withoutit we have this error : 

```
____________________________________ PipelineUserTestCase.test_get_username_no_username_in_backend ____________________________________

self = <tests.apps.social.test_pipeline_user.PipelineUserTestCase testMethod=test_get_username_no_username_in_backend>

    def setUp(self):
        """Configure social-auth strategy."""
        self.request_factory = RequestFactory()
        self.request = self.request_factory.get("/", data={"x": "1"})
>       self.strategy = load_strategy(request=self.request)

tests/apps/social/test_pipeline_user.py:21: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/local/lib/python3.7/site-packages/social_django/utils.py:23: in load_strategy
    return get_strategy(STRATEGY, STORAGE, request)
/usr/local/lib/python3.7/site-packages/social_core/utils.py:278: in get_strategy
    return Strategy(Storage, *args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <social_django.strategy.DjangoStrategy object at 0x7fc598ac7250>, storage = <class 'social_django.models.DjangoStorage'>
request = <WSGIRequest: GET '/?x=1'>, tpl = None

    def __init__(self, storage, request=None, tpl=None):
        self.request = request
>       self.session = request.session if request else {}
E       AttributeError: 'WSGIRequest' object has no attribute 'session'

/usr/local/lib/python3.7/site-packages/social_django/strategy.py:39: AttributeError
```

Comment (lunika) [reply tests/apps/social/test_pipeline_user.py:21]: But data can be removed.

**Pull request merged**